### PR TITLE
ER-108: Change Pass (Admin)

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

**Comment by saarmstrong** [comment]
I attempted a POST request and I received this error:


```
Fatal error: Uncaught Error: Class 'BasicAuthenticator' not found in /users/src/controllers/admin/UserAjax.php:490
Stack trace:
#0 /code/vendor/erdiko/core/src/Controller.php(303): erdiko\users\controllers\admin\UserAjax->postChangePass()
#1 /users/src/controllers/admin/UserAjax.php(116): erdiko\core\Controller->_autoaction('changepass', 'post')
#2 /code/vendor/erdiko/core/src/Toro.php(101): erdiko\users\controllers\admin\UserAjax->post('changepass')
#3 /code/public/default/index.php(18): Toro::serve(Array)
#4 {main}
  thrown in /users/src/controllers/admin/UserAjax.php on line 490
```


postman_and_er-108__change_pass__admin__by_jdiazarg_ _pull_request__20_ _erdiko_users_and_boing_boing

**Comment by saarmstrong** [comment]
- Do we need to log the user in to update the password on line 492? Can we just identify the user with a lookup?

**Comment by ghost** [comment]
Good comment, 
in /ajax/users/authentication/changepass we need to make sure that the change pass action is performed by someone that knows the password, then we need to log the user in and then execute the change requested. 

But, in \admin\UserAjax is a little bit different, because with each call we make an "checkAuth()" before, then we do not need log the user in to execute the change requested. So yes, is not needed.

**Comment by ghost** [comment]
Updated.

File: src/controllers/admin/UserAjax.php [hunk 1]
```diff
@@ -460,5 +460,49 @@ private function getRoleInfo($user)
...
+            // Check required fields
+            $requiredParams = array('email', 'newpass');
+            $params = (array) $data;
+            foreach ($requiredParams as $param) {
+                if (empty($params[$param])) {
+                    throw new \Exception(ucfirst($param) .' is required.');
+                }
+            }
+
+            if ($data->currentpass == $data->newpass) {
```
Comment (saarmstrong) [inline src/controllers/admin/UserAjax.php]: Does the admin really know the current password? Most likely not. 

Let's assume if the admin user is making this change, we want to set it to the 'newpass'.

File: src/controllers/admin/UserAjax.php [hunk 2]
```diff
@@ -460,5 +460,49 @@ private function getRoleInfo($user)
...
+            $data = json_decode(file_get_contents("php://input"));
+            if (empty($data)) {
+                $data = (object) $_POST;
+            }
+            // Check required fields
+            $requiredParams = array('email', 'newpass');
+            $params = (array) $data;
+            foreach ($requiredParams as $param) {
+                if (empty($params[$param])) {
+                    throw new \Exception(ucfirst($param) .' is required.');
```
Comment (saarmstrong) [inline src/controllers/admin/UserAjax.php]: let's drop the ucfirst.

File: src/controllers/admin/UserAjax.php [hunk 3]
```diff
@@ -460,5 +460,49 @@ private function getRoleInfo($user)
...
+                }
+            }
+
+            if ($data->currentpass == $data->newpass) {
+                throw new \Exception('Current pass and new pass should be different.');
+            }
+
+            $user = new User();
+            $users = $user->getByParams(array('email' => $data->email));
+            if (empty($users)) {
```
Comment (saarmstrong) [inline src/controllers/admin/UserAjax.php]: we should check to make sure the first element of the return value is actually a user instance.

**Review by saarmstrong — CHANGES_REQUESTED** [review]

**Review by saarmstrong — APPROVED** [review]

**Pull request merged**